### PR TITLE
Make all kernel magics cell magics and re-did usercodeparser

### PR DESCRIPTION
**Pull request by aggFTW**
Please focus on test_usercodeparser and understand the comments in usercodeparser.
If you want to add test cases, that'd be great.

### CODE REVIEW

**Comment by aggFTW** [comment]
I believe I made a change to one of the unit tests with the list(keys) thing that made a unit test flaky. I'm investigating.

**Comment by msftristew** [comment]
Will review tomorrow

**Comment by MohamedElKamhawy** [comment]
I don't have comments, can we merge it soon? I'm blocked on this.

File: remotespark/kernels/kernelmagics.py [hunk 1]
```diff
@@ -23,21 +23,88 @@ def __init__(self, shell, data=None):
...
+    <td>All the code in subsequent lines will be executed locally. Code must be valid Python code.</td>
+  </tr>
+</table>
+"""
+        self.ipython_display.html(help_html)
+
+    @cell_magic
+    def local(self, line, cell="", local_ns=None):
+        # This should not be reachable thanks to UserCodeParser. Registering it here so that it auto-completes with tab.
+        raise NotImplementedError("UserCodeParser should have prevented code execution from reaching here.")
```
Comment (msftristew) [inline remotespark/kernels/kernelmagics.py]: Good idea

**Comment by msftristew** [comment]
LGTM

**Pull request merged**